[Feature] 로그인 UI 구성 및 카카오 소셜 로그인 구현

stopstone opened the pull request:
## 작업 내용
카카오 소셜 로그인 구현 및 토큰 관리

## 확인 방법

```kotlin
@HiltAndroidApp
class NearApplication : Application() {
    
    override fun onCreate() {
        super.onCreate()
        
        // 카카오 SDK 초기화
        KakaoSdk.init(this, BuildConfig.KAKAO_NATIVE_APP_KEY)
        
        // 키 해시
        val keyHash = Utility.getKeyHash(this)
        Log.d("KeyHash", keyHash)
        
    }
}

```
키 해시값 확인

## 참고 사항
- 카카오 디벨로퍼스 앱에서 키 해시 넣기
- 카카오 앱 네이티브 키 local.properties 설정
<img width="485" height="1192" alt="image" src="https://github.com/user-attachments/assets/b3db7d30-4a68-43de-80bc-d91be446b650" />

<img width="484" height="1185" alt="image" src="https://github.com/user-attachments/assets/71f0f8b8-28ec-4abf-be15-bfd7a3f377a3" />



## 관련 이슈
- 토큰 만료시 토큰 재발급 하는지 확인 필요
- 카카오 로그인 앱 실행이 안되는 원인 점검

Closes #8

stopstone committed:
feat: 로그인 화면 UI 구현

- 로그인 화면 UI 구성
- 카카오 로그인 버튼 추가
- Near 로고 및 타이틀 이미지 추가
- 관련 문자열 리소스 추가
- 라이트/다크 모드에 따른 상태바 아이콘 색상 조정
- 현재는 라이트모드로 고정

stopstone committed:
feat: DataStore 라이브러리 추가

stopstone committed:
feat: 카카오 SDK 연동

stopstone committed:
feat: 로그인 화면 수정 및 네비게이션 적용

stopstone committed:
feat: 카카오 소셜 로그인 서버 API 요청

- AuthRepository 및 AuthService 인터페이스 정의
- DataStore를 사용한 토큰 관리 및 AuthRepositoryImpl 구현
- 소셜 로그인 요청/응답 데이터 클래스(SocialLoginRequest, LoginResponse, LoginResult) 추가
- 로그인 관련 ProviderType enum 정의
- LoginViewModel에 소셜 로그인 로직 및 UI 상태 관리 추가
- DI 모듈(RepositoryModule, ServiceModule)에 관련 의존성 주입 설정

stopstone committed:
feat: 카카오 로그인 서버 API 연동

- `build.gradle.kts` 파일에 카카오 네이티브 앱 키 추가
- `NearApplication.kt` 파일에 카카오 SDK 초기화 코드 추가
- `LoginScreen.kt` 파일에 카카오 로그인 로직 구현 및 UI 연동
- `AndroidManifest.xml` 파일에 카카오 로그인 콜백을 위한 Activity 및 intent-filter 추가

stopstone committed:
feat: 카카오 로그인 datasource 계층 분리 및 확장 가능한 코드로 변경

- 카카오 SDK를 사용하여 카카오 로그인 기능 구현
- SocialLoginDataSource 인터페이스 및 KakaoDataSource 구현체 추가
- SocialLoginProcessor를 통해 소셜 로그인 방식 동적 선택
- AuthRepository에 performSocialLogin 메서드 추가하여 providerType에 따라 로그인 처리 분기
- LoginViewModel에서 performLogin 메서드를 통해 소셜 로그인 요청
- LoginScreen에 카카오 로그인 버튼 추가 및 로그인 로직 연동

stopstone committed:
refactor: 소셜 로그인 데이터 소스 파일 경로 변경

stopstone committed:
Refactor: DataStoreModule 분리

DataStoreModule을 AuthRepositoryImpl에서 별도의 파일로 분리했습니다.

stopstone committed:
feat: 토큰 관리 및 인증 로직 구현

- DataStore를 사용한 토큰 저장 및 관리 클래스 (`TokenPreferences`) 추가
- 토큰 갱신 요청 및 응답 모델 (`TokenRefreshRequest`, `TokenRefreshResponse`) 추가
- `AuthRepository` 및 `AuthRepositoryImpl`에 토큰 갱신 로직 추가
- `TokenInterceptor`에서 Authorization 헤더를 동적으로 추가하고, 인증이 필요 없는 요청은 제외하도록 수정
- `LoginResponse` 모델에서 `refreshTokenInfo`를 `refreshToken`으로 변경
- `AuthService`에 토큰 갱신 API (`/auth/renew`) 인터페이스 추가

stopstone committed:
feat: 토큰 갱신 및 만료 처리 로직 개선

- `AuthRepositoryImpl`
    - 소셜 로그인 시 만료 시간 계산 및 저장 로직 추가
    - 토큰 갱신 재시도 로직 (`refreshTokenWithRetry`) 추가
    - 401 에러 메시지 수정

- `TokenPreferences`
    - 토큰 만료 시간 (`expires_at`) 저장 및 조회 기능 추가
    - 토큰 만료 여부 확인 (`isTokenExpired`) 기능 추가
    - 액세스 토큰 관찰 (`observeAccessToken`) 기능 추가
    - 기존 `saveAccessToken`, `saveRefreshToken` 메서드 삭제

- `TokenRefreshResponse`, `LoginResponse`
    - `refreshToken` 필드를 `refreshTokenInfo` (객체)로 변경하여 만료 시간 정보 포함

- `TokenInterceptor`
    - 자동 토큰 갱신 및 재시도 로직 구현
    - 토큰 만료 시간 기반의 유효성 검사 로직 추가
    - 토큰 변화를 관찰하여 실시간으로 `currentToken`, `tokenExpiresAt` 업데이트
    - 인증 제외 경로를 `AuthEndpoint` enum으로 관리
    - 토큰 갱신 실패 시 토큰 삭제 처리

stopstone committed:
refactor: 토큰 갱신 로직 삭제

stopstone committed:
refactor: 로그인 백그라운드 색상 변경

stopstone committed:
chore: 카카오 네이티브키 CI 설정

stopstone committed:
feat: 토큰 갱신 로직 추가

- `TokenAuthenticator`를 구현하여 401 에러 발생 시 토큰 갱신 및 재요청 로직 추가
- `TokenInterceptor`에서 토큰 만료 처리 로직 제거 (Authenticator에서 처리)
- `OkHttpClient`에 `TokenAuthenticator` 적용

stopstone committed:
feat: TokenManager 추가 및 토큰 관리 로직 개선

- `TokenManager` 클래스를 추가하여 토큰 저장, 조회, 갱신 로직을 중앙에서 관리합니다.
- `TokenAuthenticator`, `AuthRepositoryImpl`, `TokenInterceptor`에서 `TokenManager`를 사용하도록 수정하여 토큰 관리 책임을 위임합니다.
- `AuthRepositoryImpl`에서 중복된 토큰 관련 로직 (만료 시간 계산 등)을 제거하고 `TokenManager`의 기능을 사용합니다.
- `TokenAuthenticator`에서 `runBlocking`을 사용하여 `suspend` 함수를 호출하도록 수정했습니다.
- `TokenInterceptor`에서 `runBlocking`을 사용하여 토큰을 가져오고, 토큰 변화를 관찰하는 로직을 제거하여 `TokenManager`에 의존하도록 단순화했습니다.

stopstone committed:
refactor: 카카오 로그인 로직 수정

카카오 로그인 시 context를 전역으로 참조하도록 변경

stopstone committed:
refactor: 토큰 유효성 검사 로직 수정

- DataStore에서 직접 토큰 및 만료 시간을 가져와서 유효성 검사
- `isTokenExpired()` 메서드 호출 제거 및 로직 통합

stopstone committed:
refactor: logout 로직 수정

stopstone committed:
refactor: 토큰 만료 시간 계산 로직 변경

- SimpleDateFormat에서 java.time.LocalDateTime으로 변경하여 스레드 안전성 확보
- DateTimeParseException 발생 시 null 반환하도록 수정

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/datasource/KakaoDataSource.kt:
```diff
@@ -0,0 +1,89 @@
+        override val supportedType: ProviderType = ProviderType.KAKAO
+
+        override suspend fun login(): Result<String> =
+            try {
```
try catch로 꼼꼼하게 예외 처리 주셨군요!
kotlin의 runCatching을 통해 간결하게 표현해보면 어떨까요?
Result로 값을 리턴하는 부분에서도 이점이 있을 것 같아요!

stopstone replied:
맞네요!
돌아가는거만 신경쓰다가.. 간결한 코드에 대해서 신경을 못 쓴 것 같습니다
캐치해주셔서 감사드려요!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/model/LoginResult.kt:
```diff
@@ -0,0 +1,8 @@
+
+data class LoginResult(
+    val isSuccess: Boolean,
+    val accessToken: String? = null,
```
뷰로 보내는 모델로 보입니다! accessToken과 refreshToken이 포함되어야 할까요?

stopstone replied:
불필요 하겠군요! 감사합니다!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/network/auth/TokenInterceptor.kt:
```diff
@@ -1,21 +1,64 @@
+    }
+
+    companion object {
+        enum class AuthEndpoint(val path: String) {
```
companion object 내에서 enum class로 선언주신 이유가 있을까요?

stopstone replied:
**1. enum으로 작성한 이유**
우선 헤더가 필요없는 인증 관련 엔트포인트를 enum으로 처리하고 싶었습니다!
그렇지 않으면 .contains.contains로 점차 늘어날 것 같아서요..!

**2. companion 내부에 작성한 이유**
그렇다고 model이나 다른 파일로 분리하기엔 아직까지는 내부에서만 활용되는 상수라고 판단하여
companion 안에 넣어두었습니다.

**코멘트**
외부에서 활용될 수 있도록 분리하는 것이 좋을까요?
이 부분에 대해서는 감이 잘 서지 않네요..!

rhkrwngud445 replied:
답변 감사합니다! 
저는 분리되는게 좋은 것 같아요! 추후 service단의 엔드포인트를 enum으로 관리할 때 재활용성이나 관리측면에서 이점이 있을 것 같아요!

추가적으로 enum이 싱글톤으로 동작하다보니, companion object 내의 선언이 의미가 모호해질 수 있을 것 같아요!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/network/auth/TokenInterceptor.kt:
```diff
@@ -1,21 +1,64 @@
+            TOKEN_RENEW("/auth/renew");
+
+            companion object {
+                val EXCLUDED_PATHS = listOf(
```
enum의 entries 속성을 사용하면 리스트를 뽑아올 수 있습니다!
AuthEndpoint.entries 형식으로요!

stopstone replied:
감사합니다!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/datasource/KakaoDataSource.kt:
```diff
@@ -0,0 +1,89 @@
+                    when {
+                        error != null -> {
+                            if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
+                                continuation.resume("")
```
nullable로 처리하는 것은 어떨까요?

stopstone replied:
오! 고려 해보겠습니다 :D

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/di/RepositoryModule.kt:
```diff
@@ -20,4 +25,12 @@ interface RepositoryModule {
+
+    @Binds
+    @IntoSet
+    abstract fun bindKakaoDataSource(kakaoDataSource: KakaoDataSource): SocialLoginDataSource
```
repository와 DataSource는 같은 추상화 레벨로 보기 어려워 보입니다!
별도의 DI로 구성하는 것은 어떨까요?

stopstone replied:
맞는 말씀입니다! 이건 바로 수정할게요!
코드를 여기저기 옮겨쓰다보니... 꼼꼼하지 못했던 것 같아요 감사합니다!!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/data/datasource/SocialLoginProcessor.kt:
```diff
@@ -0,0 +1,32 @@
+class SocialLoginProcessor
+    @Inject
+    constructor(
+        private val socialLoginDataSources: Set<@JvmSuppressWildcards SocialLoginDataSource>,
```
Set과 JvmSuppressWildcards를 적용주신 이유가 궁금합니다!

stopstone replied:
**1. Set을 적용한 이유**
현재는 카카오 로그인 뿐이지만 추후 다른 소셜 로그인을 확장하기 쉽게 socialLoginDataSources를 구성했습니다!
이때 List보다는 중복을 방지할 수 있다는 차원에서도 Set이 명확할 것 같아 Set으로 반환하였습니다

**2. JvmSuppressWildcards를 적용한 이유**
// Kotlin에서 이렇게 쓰면
val dataSources: Set<SocialLoginDataSource>

// JVM 바이트코드에서는 이렇게 변환돼요
Set<? extends SocialLoginDataSource>

여기서 ?는 wildcard로, "정확한 타입을 모르지만 SocialLoginDataSource의 하위 타입들"이라는 의미입니다. 
이는 Kotlin의 타입 안전성을 위한 공변성 때문인데, 
Dagger와 Hilt는 Java로 작성된 라이브러리라서 이런 wildcard 타입을 제대로 처리하지 못합니다. 

바인딩할 때 "어? 이게 정확히 어떤 타입인지 모르겠네?" 하면서 에러가 발생할 가능성이 있어,
@JvmSuppressWildcards 어노테이션을 사용해 wildcard 생성을 억제하여 Set<SocialLoginDataSource>로 그대로 변환되도록 했습니다. 
이렇게 하면 Dagger가 정확한 타입을 인식하고 올바르게 의존성 주입을 할 수 있어요
어노테이션을 사용하지 않으면 의존성 에러가 발생하더라구요!!

rhkrwngud445 replied:
이해가 쏙쏙 되네요! 상세한 답변 감사합니다!

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/network/auth/TokenManager.kt:
```diff
@@ -0,0 +1,125 @@
+@Inject
+constructor(
+    private val tokenPreferences: TokenPreferences,
+    private val authServiceProvider: Provider<AuthService>, // Provider로 지연 주입
```
Provider로 순횐참조를 해결할 수 있군요! 배워갑니다!

stopstone replied:
저도 이번에 알게 되었어요! 
지연 주입이 가능하다고 하는데, 어떤 이점이 더 있는지는 더 알아봐야 겠습니다!! :D

rhkrwngud445 reviewed (commented):
첫 PR 고생하셨습니다!👍
궁금한 점과 개선해볼만한 점 위주로 남겼습니다!
수정 대신 반박 코멘트도 좋습니다 차차 코드 스타일 맞춰가요😀

stopstone committed:
refactor: 소셜 로그인 결과 처리 방식 변경

- LoginResult 모델을 제거하고 Result<Unit>을 사용하도록 수정

stopstone committed:
Refactor: TokenInterceptor에서 AuthEndpoint 분리

- 기존 TokenInterceptor 내부에 있던 AuthEndpoint enum을 별도의 파일로 분리하여 관리하도록 수정했습니다.
- 이를 통해 TokenInterceptor는 인증 로직에만 집중하고, 인증 예외 경로는 AuthEndpoint enum에서 관리하도록 역할을 분리했습니다.

stopstone committed:
refactor: KakaoDataSource 바인딩 DI 모듈 뷰ㅜㄴ리

KakaoDataSource를 SocialLoginDataSource로 바인딩하는 로직을 RepositoryModule에서 DataSourceModule로 이동했습니다.
